File: Day13/day13.py
import sys


def preprocess_lists(read_data: str):
    read_data = read_data.removesuffix("\n")
    read_data = read_data.split("\n\n")
    read_data = [lists.split("\n") for lists in read_data]
    parsed_lists = []

    for pair_of_lists in read_data:
        parsed_lists.append([eval(pair_of_lists[0]), eval(pair_of_lists[1])])

    return parsed_lists


def preprocess_all_lists(read_data: str):
    read_data = read_data.replace("\n\n", "\n")
    read_data = read_data.split("\n")
    read_data.remove("")
    parsed_lists = [eval(line) for line in read_data]

    return parsed_lists


def compare_lists(list_1: list, list_2: list):
    value = compare_sublists(list_1, list_2)
    if value == "continue":
        if len(list_1) == 0:
            return True
        return compare_lists(list_1[1:], list_2[1:])
    return value


def compare_sublists(list_1: list, list_2: list):
    if len(list_1) == 0:
        if len(list_2) > 0:
            return True
        return "continue"
    if len(list_2) == 0:
        return False

    # The case where they both are lists
    if isinstance(list_1[0], int) and isinstance(list_2[0], int):
        if list_1[0] < list_2[0]:
            return True
        if list_1[0] > list_2[0]:
            return False
        return compare_sublists(list_1[1:], list_2[1:])

    # The case where one or both are lists.
    # First we convert the next int to list, if needed.
    if isinstance(list_1[0], list) and isinstance(list_2[0], int):
        list_1_to_compare = list_1[0]
        list_2_to_compare = [list_2[0]]
    elif isinstance(list_2[0], list) and isinstance(list_1[0], int):
        list_1_to_compare = [list_1[0]]
        list_2_to_compare = list_2[0]
    else:
        list_1_to_compare = list_1[0]
        list_2_to_compare = list_2[0]

    # We compare the sublists.
    value = compare_sublists(list_1_to_compare, list_2_to_compare)
    if value == "continue":
        return compare_sublists(list_1[1:], list_2[1:])
    return value


def decoder_key(lists_to_sort):
    start_packet = [[2]]
    end_packet = [[6]]

    index_start = 1
    index_end = 2

    for packet in lists_to_sort:
        if compare_lists(packet, start_packet):
            index_start += 1
            index_end += 1
        elif compare_lists(packet, end_packet):
            index_end += 1

    return index_start * index_end


if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
    except IndexError:
        file_name = "input.txt"
    lists_to_compare = open(file_name).read()
    parsed_lists = preprocess_lists(lists_to_compare)
    in_order = 0
    for index, pair_of_lists in enumerate(parsed_lists):
        if compare_lists(pair_of_lists[0], pair_of_lists[1]):
            in_order += index + 1

    lists_to_order = preprocess_all_lists(lists_to_compare)

    print(f"The sum of indices of lists in order is {in_order}.")
    print(f"The decoder key is {decoder_key(lists_to_order)}.")

File: Day13/day13.rs
use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
use std::str::from_utf8;

use Packet::*;

fn get_data() -> String {
    let input_args: Vec<String> = std::env::args().collect();
    let mut file_name: String = "input.txt".to_string();
    if input_args.len() > 1 {
        file_name = input_args[1].clone();
    }
    let mut data = std::fs::read_to_string(file_name).unwrap();
    if data.ends_with("\n") {
        data.pop();
    }
    return data;
}


fn pairs_in_order(packets: &Vec<Packet>) -> usize {
    let mut answer = 0;

    packets.chunks(2).enumerate().for_each(|(index, pair)| {
        let first = &pair[0];
        let second = &pair[1];
        if first <= second {
            answer += index + 1;
        }
    });
    answer
}

fn decoder_key(packets: &Vec<Packet>) -> usize {

    let start_packet = parse_packet("[[2]]".as_bytes(), &mut 0);
    let end_packet = parse_packet("[[6]]".as_bytes(), &mut 0);

    let mut index_start = 1;
    let mut index_end = 2;

    packets.into_iter().for_each(|packet| {
        if packet <= &start_packet {
            index_start += 1;
            index_end += 1;
        } else if packet <= &end_packet {
            index_end += 1;
        }
    });
    index_start * index_end
}


#[derive(Clone, PartialEq, Eq)]
enum Packet {
    Int(usize),
    List(Vec<Self>),
}

fn parse_packet(string_line: &[u8], index: &mut usize) -> Packet {
    let mut packets = Vec::new();

    *index += 1;
    while string_line[*index] != b']' {
        if string_line[*index].is_ascii_digit() {
            let start = *index;
            while string_line[*index].is_ascii_digit() {
                *index += 1;
            }

            let val: usize = from_utf8(&string_line[start..*index]).unwrap().parse().unwrap();
            packets.push(Int(val))
        } else if string_line[*index] == b',' {
            *index += 1;
        } else if string_line[*index] == b'[' {
            packets.push(parse_packet(string_line, index));
        }
    }
    *index += 1;

    List(packets)
}

fn preprocess_input(input: String) -> Vec<Packet> {
    let mut packets: Vec<Packet> = vec![];
    input.replace("\n\n", "\n").split("\n").for_each(|line| {
        packets.push(parse_packet(&line.as_bytes(), &mut 0))
    });
    packets
}

impl Ord for Packet {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Int(v1), Int(v2)) => v1.cmp(v2),
            (Int(v), List(_)) => List(vec![Int(*v)]).cmp(other),
            (List(_), Int(v)) => self.cmp(&List(vec![Int(*v)])),
            (List(packets1), List(packets2)) => packets1.cmp(packets2),
        }
    }
}

impl PartialOrd for Packet {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn main() {
    let input = get_data();
    let packets = preprocess_input(input);
    let in_order = pairs_in_order(&packets);
    let key = decoder_key(&packets);
    println!("The sum of indices of lists in order is {}", in_order);
    println!("The decoder key is {}.", key)
}
